server/sonar-process-monitor/src/main/java/org/sonar/application/SchedulerImpl.java

```diff
@@ -0,0 +1,278 @@
+import org.sonar.application.config.AppSettings;
+import org.sonar.application.config.AppSettingsHelper;
+import org.sonar.application.process.JavaCommand;
+import org.sonar.application.process.JavaCommandFactory;
+import org.sonar.application.process.JavaProcessLauncher;
+import org.sonar.application.process.Lifecycle;
+import org.sonar.application.process.ProcessEventListener;
+import org.sonar.application.process.ProcessLifecycleListener;
+import org.sonar.application.process.SQProcess;
+import org.sonar.process.ProcessId;
+
+public class SchedulerImpl implements Scheduler, ProcessEventListener, ProcessLifecycleListener, AppStateListener {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SchedulerImpl.class);
+
+  private final AppSettings settings;
+  private final AppReloader appReloader;
+  private final JavaCommandFactory javaCommandFactory;
+  private final JavaProcessLauncher javaProcessLauncher;
+  private final AppState appState;
+  private final NodeLifecycle nodeLifecycle = new NodeLifecycle();
+
+  private final CountDownLatch keepAlive = new CountDownLatch(1);
+  private final AtomicBoolean restartRequested = new AtomicBoolean(false);
+  private final AtomicBoolean restartDisabled = new AtomicBoolean(false);
+  private final EnumMap<ProcessId, SQProcess> processesById = new EnumMap<>(ProcessId.class);
+  private final AtomicInteger operationalCountDown = new AtomicInteger();
+  private final AtomicInteger stopCountDown = new AtomicInteger();
+  private StopperThread stopperThread;
+  private RestarterThread restarterThread;
+  private long processWatcherDelayMs = SQProcess.DEFAULT_WATCHER_DELAY_MS;
+
+  public SchedulerImpl(AppSettings settings, AppReloader appReloader, JavaCommandFactory javaCommandFactory,
+    JavaProcessLauncher javaProcessLauncher,
+    AppState appState) {
+    this.settings = settings;
+    this.appReloader = appReloader;
+    this.javaCommandFactory = javaCommandFactory;
+    this.javaProcessLauncher = javaProcessLauncher;
+    this.appState = appState;
+    this.appState.addListener(this);
+  }
+
+  SchedulerImpl setProcessWatcherDelayMs(long l) {
+    this.processWatcherDelayMs = l;
+    return this;
+  }
+
+  @Override
+  public void schedule() {
+    if (!nodeLifecycle.tryToMoveTo(NodeLifecycle.State.STARTING)) {
+      return;
+    }
+    processesById.clear();
+
+    for (ProcessId processId : AppSettingsHelper.getEnabledProcesses(settings)) {
+      SQProcess process = SQProcess.builder(processId)
+        .addProcessLifecycleListener(this)
+        .addEventListener(this)
+        .setWatcherDelayMs(processWatcherDelayMs)
+        .build();
+      processesById.put(process.getProcessId(), process);
+    }
+    operationalCountDown.set(processesById.size());
+    stopCountDown.set(processesById.size());
+
+    tryToStartAll();
+  }
+
+  private void tryToStartAll() {
+    try {
+      tryToStartEs();
+      tryToStartWeb();
+      tryToStartCe();
+    } catch (RuntimeException e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)